packages/loader/container-loader/src/containerStorageAdapter.ts

```diff
@@ -0,0 +1,94 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+    IDocumentStorageService,
+    IDocumentStorageServicePolicies,
+    ISummaryContext,
+} from "@fluidframework/driver-definitions";
+import {
+    ICreateBlobResponse,
+    ISnapshotTree,
+    ISummaryHandle,
+    ISummaryTree,
+    ITree,
+    IVersion,
+} from "@fluidframework/protocol-definitions";
+import { IDisposable } from "@fluidframework/common-definitions";
+import { AttachState } from "@fluidframework/container-definitions";
+
+/**
+ * This class wraps the actual storage and make sure no wrong apis are called according to
+ * container attach state.
+ */
+export class ContainerStorageAdapter implements IDocumentStorageService, IDisposable {
+    private _disposed = false;
+    constructor(
+        private readonly storageGetter: () => IDocumentStorageService,
+        private readonly attachState: () => AttachState,
+        private readonly blobs: Map<string, ArrayBufferLike>,
+    ) {
+    }
+
+    private throwOnNotAttached(name: string) {
+        const attachState = this.attachState();
+        if (attachState !== AttachState.Attached) {
+            throw new Error(`${name} not allowed in Unattached container`);
+        }
+    }
+
+    public get policies(): IDocumentStorageServicePolicies | undefined {
+        return this.storageGetter().policies;
+    }
+
+    public get disposed() {return this._disposed;}
+    public dispose() {
+        this._disposed = true;
```

Comment:
Do we need to propagate call to inner storage?

Reply:
Actually figured out that this should not implement IDisposable as we just pass IDocumentStorageService to lower layers and only contaihner closes storage on close which it will still do with internal actual storage.